File: dp/53.go
func maxSubArray(nums []int) int {
    if len(nums) == 0 {
        return 0
    }
    var ret, cur int
    cur = nums[0]
    ret = nums[0]
    for i := 1; i < len(nums); i++ {
        if nums[i] + cur < nums[i] {
            cur = nums[i]
        }else {
            cur = nums[i] + cur
        }
        if ret < cur {
            ret = cur
        }
    }
    return ret
}

File: dp/53.cc
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        if (nums.empty()) {
            return 0;
        }
        int ret = INT_MIN;
        int cur = 0;
        for (int i = 0; i < nums.size(); i++) {
            if (cur < 0) {
                cur = nums[i];
            } else {
                cur += nums[i];
            }
            ret = cur > ret ? cur : ret;            
        }
        return ret;
    }
};
